File: api/v1/schemas/visualization.py
# -*- coding: utf-8 -*-
"""
===================================
可视化数据相关模型
===================================

职责：
1. 定义可视化相关的请求和响应模型
"""

from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field


class VisualizationIndicatorData(BaseModel):
    """单个指标数据"""
    
    indicator_type: str = Field(..., description="指标类型")
    data: List[Dict[str, Any]] = Field(..., description="指标数据列表")


class VisualizationResponse(BaseModel):
    """可视化数据响应"""
    
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    kline_data: List[Dict[str, Any]] = Field(..., description="K线数据")
    indicators: List[VisualizationIndicatorData] = Field(default_factory=list, description="指标数据列表")
    chip_distribution: Optional[Dict[str, Any]] = Field(None, description="筹码分布数据")
    circulating_shares_updated: bool = Field(False, description="流通股本是否已更新")
    circulating_shares: Optional[float] = Field(None, description="最新流通股本")
    turnover_filled_count: int = Field(0, description="填充的历史换手率记录数")
    
    class Config:
        json_schema_extra = {
            "example": {
                "stock_code": "600519",
                "stock_name": "贵州茅台",
                "kline_data": [],
                "indicators": []
            }
        }


class VisualizationSearchHistoryItem(BaseModel):
    """可视化搜索历史项"""
    
    id: int = Field(..., description="记录ID")
    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    searched_at: str = Field(..., description="搜索时间")
    selected_indicators: List[str] = Field(default_factory=list, description="选中的指标列表")


class VisualizationSearchHistoryResponse(BaseModel):
    """可视化搜索历史响应"""
    
    items: List[VisualizationSearchHistoryItem] = Field(default_factory=list, description="搜索历史列表")
    
    class Config:
        json_schema_extra = {
            "example": {
                "items": []
            }
        }


class ChipDistributionResponse(BaseModel):
    """筹码分布响应"""
    
    stock_code: str = Field(..., description="股票代码")
    price_bins: List[float] = Field(..., description="价格区间列表")
    chip_volumes: List[float] = Field(..., description="筹码数量列表")
    profit_volumes: List[float] = Field(..., description="获利盘数量列表")
    loss_volumes: List[float] = Field(..., description="套牢盘数量列表")
    avg_cost: Optional[float] = Field(None, description="平均成本价")
    max_chip_price: Optional[float] = Field(None, description="筹码集中度最高价格")
    current_price: Optional[float] = Field(None, description="当前价格")


class VisualizationSearchRequest(BaseModel):
    """可视化搜索请求"""

    stock_code: str = Field(..., description="股票代码")
    stock_name: Optional[str] = Field(None, description="股票名称")
    days: int = Field(60, ge=1, le=3650, description="获取天数")
    indicator_types: Optional[List[str]] = Field(None, description="指标类型列表，不填则返回所有可用指标")

File: apps/dsa-web/src/api/visualization.ts
import apiClient from './index';

export type KLineDataPoint = {
  date: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
  amount?: number;
  pct_chg?: number;
};

export type IndicatorDataPoint = {
  [key: string]: any;
  date?: string;
};

export type IndicatorData = {
  indicator_type: string;
  data: IndicatorDataPoint[];
  metadata?: {
    main_net_inflow?: number;
    super_net_inflow?: number;
    big_net_inflow?: number;
    medium_net_inflow?: number;
    small_net_inflow?: number;
    [key: string]: any;
  };
};

export type VisualizationResponse = {
  stock_code: string;
  stock_name?: string;
  kline_data: KLineDataPoint[];
  indicators: IndicatorData[];
  chip_distribution?: {
    latest?: ChipDistributionResponse;
    history?: Array<{ date: string; data: ChipDistributionResponse }>;
  };
};

export type VisualizationSearchHistoryItem = {
  id: number;
  stock_code: string;
  stock_name?: string;
  searched_at: string;
  selected_indicators: string[];
  days?: number;
};

export type VisualizationSearchHistoryResponse = {
  items: VisualizationSearchHistoryItem[];
};

export type ChipDistributionResponse = {
  stock_code: string;
  price_bins: number[];
  chip_volumes: number[];
  profit_volumes: number[];
  loss_volumes: number[];
  avg_cost?: number;
  max_chip_price?: number;
  current_price?: number;
};

export type VisualizationSearchRequest = {
  stock_code: string;
  stock_name?: string;
  days?: number;
  indicator_types?: string[];
};

export const visualizationApi = {
  async getVisualizationData(
    stockCode: string,
    days: number = 3650,
    indicatorTypes?: string[],
    startDate?: string
  ): Promise<VisualizationResponse> {
    const params: Record<string, any> = {};
    if (startDate) {
      params.start_date = startDate;
    } else {
      params.days = days;
    }
    if (indicatorTypes && indicatorTypes.length > 0) {
      params.indicator_types = indicatorTypes.join(',');
    }

    const response = await apiClient.get(`/api/v1/visualization/${encodeURIComponent(stockCode)}`, {
      params,
    });

    return response.data as VisualizationResponse;
  },

  async getSearchHistory(limit: number = 100): Promise<VisualizationSearchHistoryResponse> {
    const response = await apiClient.get('/api/v1/visualization/history', {
      params: { limit },
    });

    return response.data as VisualizationSearchHistoryResponse;
  },

  async saveSearchHistory(request: VisualizationSearchRequest): Promise<VisualizationSearchHistoryItem> {
    const response = await apiClient.post('/api/v1/visualization/history', request);

    return response.data as VisualizationSearchHistoryItem;
  },

  async deleteSearchHistory(recordId: number): Promise<void> {
    await apiClient.delete(`/api/v1/visualization/history/${recordId}`);
  },

  async updateSearchHistoryTimestamp(recordId: number): Promise<VisualizationSearchHistoryItem> {
    const response = await apiClient.patch(`/api/v1/visualization/history/${recordId}/timestamp`);
    return response.data as VisualizationSearchHistoryItem;
  },

  async getChipDistribution(
    stockCode: string,
    days: number = 365,
    startDate?: string,
    endDate?: string,
    endDateIdx?: number
  ): Promise<ChipDistributionResponse> {
    const params: Record<string, any> = { days };
    if (startDate) {
      params.start_date = startDate;
    }
    if (endDate) {
      params.end_date = endDate;
    }
    if (endDateIdx !== undefined) {
      params.end_date_idx = endDateIdx;
    }

    const response = await apiClient.get(`/api/v1/visualization/chip-distribution/${encodeURIComponent(stockCode)}`, {
      params,
    });

    return response.data as ChipDistributionResponse;
  },
};
